ons-spark/ancillary-topics/pandas_udfs_rescue.ipynb

```diff
@@ -0,0 +1,902 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "# Pandas UDFs"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Generally you will want to try and use native Spark functions wherever possible as this will run faster than Python code. However, there are instances where there may not be a Spark function available to perform what you need. In cases such as this, the second best option is to use a Pandas UDF. \n",
+    "\n",
+    "### Why are Pandas UDFs better than Python UDFs?\n",
+    "\n",
+    "Normal Python UDFs are not efficient in Spark as they do not work on the Spark cluster and can only be processed in the Python runtime. As the code for Python UDFs cannot be executed in the Java Virtual Machine (JVM), the platform which runs Spark, each row of the dataframe is serialised and deserialised between the Python runtime and the JVM. As can be imagined, this causes massive (de)serialisation overheads, high data copy in memory and is very slow!\n",
+    "\n",
+    "Pandas UDFs (also known as vectorised UDFs) can work on the Spark executors to process data in a distributed manner and allow for vectorised operations. This means that Pandas UDFs can work on the whole dataframe at once instead of just one row at a time like Python UDFs. This vectorisation is achieved by using Apache Arrow to transfer data across the cluster between the JVM and the Python executors with very low data copy in memory and (de)serialisation overheads. Essentially, Apache Arrow acts as a middleman to store a single copy of the data which can be accessed by both Python and Java processes.\n",
```

Comment:
I'm not sure that Pandas UDFs can work on a whole dataframe at a time. Is it that they can work on a partition at a time.